cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,1209 @@
+    case Type::INT16:
+      return visitor.Visit(static_cast<Int16Builder*>(builder));
+    case Type::UINT32:
+      return visitor.Visit(static_cast<UInt32Builder*>(builder));
+    case Type::INT32:
+      return visitor.Visit(static_cast<Int32Builder*>(builder));
+    case Type::UINT64:
+      return visitor.Visit(static_cast<UInt64Builder*>(builder));
+    case Type::INT64:
+      return visitor.Visit(static_cast<Int64Builder*>(builder));
+    case Type::HALF_FLOAT:
+      return visitor.Visit(static_cast<HalfFloatBuilder*>(builder));
+    case Type::FLOAT:
+      return visitor.Visit(static_cast<FloatBuilder*>(builder));
+    case Type::DOUBLE:
+      return visitor.Visit(static_cast<DoubleBuilder*>(builder));
+    case Type::STRING:
+      return visitor.Visit(static_cast<StringBuilder*>(builder));
+    case Type::BINARY:
+      return visitor.Visit(static_cast<BinaryBuilder*>(builder));
+    case Type::FIXED_SIZE_BINARY:
+      return visitor.Visit(static_cast<FixedSizeBinaryBuilder*>(builder));
+    case Type::DATE32:
+      return visitor.Visit(static_cast<Date32Builder*>(builder));
+    case Type::DATE64:
+      return visitor.Visit(static_cast<Date64Builder*>(builder));
+    case Type::TIMESTAMP:
+      return visitor.Visit(static_cast<TimestampBuilder*>(builder));
+    case Type::TIME32:
+      return visitor.Visit(static_cast<Time32Builder*>(builder));
+    case Type::TIME64:
+      return visitor.Visit(static_cast<Time64Builder*>(builder));
+    case Type::INTERVAL:
+      return Status::NotImplemented("No IntervalBuilder");
+    case Type::DECIMAL:
+      return visitor.Visit(static_cast<Decimal128Builder*>(builder));
+    case Type::LIST:
+      return visitor.Visit(static_cast<ListBuilder*>(builder));
+    case Type::STRUCT:
+      return visitor.Visit(static_cast<StructBuilder*>(builder));
+    case Type::UNION:
+      return Status::NotImplemented("No UnionBuilder");
+    case Type::DICTIONARY:
+      return Status::NotImplemented("No DictionaryBuilder");
+    case Type::MAP:
+      return Status::NotImplemented("No MapBuilder");
+    default:
+      return Status::NotImplemented("Unknown Builder");
+  }
+}
+
+class TypedHandler
+    : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, TypedHandler> {
+ public:
+  TypedHandler(StructBuilder* root_builder) : builder_(root_builder) {}
+
+  bool Null() {
+    if (Skipping()) return true;
+    // TODO check whether we're currently nullable
+    status_ = VisitBuilder(builder_, AppendNullVisitor{});
+    return status_.ok();
+  }
+
+  bool Bool(bool value) {
+    if (Skipping()) return true;
+    if (builder_->type()->id() == Type::BOOL) {
+      status_ = static_cast<BooleanBuilder*>(builder_)->Append(value);
+    } else {
+      status_ = ConversionError();
+    }
+    return status_.ok();
+  }
+
+  bool RawNumber(const char* data, rapidjson::SizeType size, bool) {
+    if (Skipping()) return true;
+    status_ = VisitBuilder(builder_, AppendNumberVisitor{data, size});
```

Comment:
In this case, `VisitBuilder(builder_, AppendNumberVisitor{data, size})` evaluates to a simple switch: if the builder has type INT64 or DOUBLE, then append. If the builder is NULL, then promote to INT64 and append. Otherwise emit a conversion error. I'm not sure that amount of dynamic dispatch can be avoided.

For consistency with the rest of our parsing, I definitely want to use StringConverter rather than rely on RapidJSON's conversions. I'm not sure what you mean by slow path, though

Reply:
Oh wait, that's not the inferring handler

Reply:
In any case, the dynamic dispatch will be removed from SAX handling when I no longer do conversions there. Doing this sort of dynamic dispatch at the conversion stage is something I'd like to talk more about; CSV avoids it by visiting a column during conversion. [See related discussion](https://github.com/apache/arrow/pull/2576#discussion_r221577736)